### PR TITLE
Automatic bisecting script

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Automatically identifies a commit based on whether or not it compiles, or whether or not its emit contains a particular string.

Usage:
```
C:\github\TypeScript-dev>type test.ts
declare function doFoo(bar?: { baz?: string; bar: number; }): string;
declare function doFoo(bar?: any): number;

var x: string = doFoo({ bar: 1 }); // x: number, should be x: string

C:\github\TypeScript-dev>scripts\bisect HEAD v1.3 test.ts compiles
[ ... lots of output ... ]
81eaae0afa37ff150456ccc058a153ebb0f0c428 is the first bad commit
commit 81eaae0afa37ff150456ccc058a153ebb0f0c428
[ ... commit metadata ...]
bisect run success
```

### CODE REVIEW

File: scripts/bisect-test.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/// <reference path="..\src\harness\external\node.d.ts" />
+
+import cp = require('child_process');
+import fs = require('fs');
+
+var args = process.argv.slice(2);
+
+var jake = cp.exec('jake clean local', () => void 0);
+jake.on('close', code => {
```
Comment (JsonFreeman) [inline scripts/bisect-test.ts]: I like exitCode

File: scripts/bisect-test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+var args = process.argv.slice(2);
+
+var jake = cp.exec('jake clean local', () => void 0);
+jake.on('close', code => {
+    if (code === 0) {
+        // See what we're being asked to do
+        if (args[1] === 'compiles' || args[1] === '!compiles') {
+            var tsc = cp.exec('node built/local/tsc.js ' + args[0], () => void 0);
+            tsc.on('close', tscCode => {
+                if ((tscCode === 0) === (args[1] === 'compiles')) {
```
Comment (JsonFreeman) [inline scripts/bisect-test.ts]: If the equalities are equal... haha!

File: scripts/bisect-test.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            });
+        } else {
+            console.log('Unknown command line arguments.');
+            console.log('Usage (compile errors): git bisect run scripts\bisect.js "foo.ts --module amd" compiles');
+            console.log('Usage (emit check): git bisect run scripts\bisect.js bar.ts emits bar.js "_this = this"');
+            process.exit(-1);
+        }
+    } else {
+        // Build failed
+        process.exit(125); // bisect skip
```
Comment (JsonFreeman) [inline scripts/bisect-test.ts]: What does skipping do?

File: scripts/bisect-test.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                        console.log('Bad');
+                        process.exit(1); // Bad
+                    }
+                });
+            });
+        } else {
+            console.log('Unknown command line arguments.');
+            console.log('Usage (compile errors): git bisect run scripts\bisect.js "foo.ts --module amd" compiles');
+            console.log('Usage (emit check): git bisect run scripts\bisect.js bar.ts emits bar.js "_this = this"');
+            process.exit(-1);
```
Comment (JsonFreeman) [inline scripts/bisect-test.ts]: How does bisect interpret -1?

**Comment by JsonFreeman** [comment]
This is awesome. Now we can have an extensible script for bisecting, and we can add things like declaration emit, and such.

File: scripts/bisect-test.ts (hunk 3 above)
Comment (RyanCavanaugh) [inline scripts/bisect-test.ts]: It tells `git bisect` that the current commit is not testable (for example, the compiler doesn't build)

File: scripts/bisect-test.ts (hunk 4 above)
Comment (RyanCavanaugh) [inline scripts/bisect-test.ts]: "abort"

File: scripts/bisect-test.ts (hunk 3 above)
Comment (yuit) [inline scripts/bisect-test.ts]: In that case would it print out an error message?

File: scripts/bisect-test.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline scripts/bisect-test.ts]: How unexpressive - this is why we need to support the boolean NXOR operator `!^^`.

File: scripts/bisect-test.ts (hunk 3 above)
Comment (RyanCavanaugh) [inline scripts/bisect-test.ts]: It's basically "unknown" rather than pass/fail; git bisect then tries a different commit to continue the bisect operation

**Comment by DickvdBrink** [comment]
Cool script! :+1:

**Pull request merged**